config/operation.go

```diff
@@ -0,0 +1,496 @@
+	if err != nil {
+		return err
+	}
+
+	c, save, err := ReadWriteConfig()
+	if err != nil {
+		return err
+	}
+
+	c.Url = verified
+
+	return save()
+}
+
+func AddUser(rawText string) error {
+	c, save, err := ReadWriteConfig()
+	if err != nil {
+		return err
+	}
+
+	list := strings.Split(rawText, ",")
+	for _, input := range list {
+		email, err := verifyEmail(input)
+		if err != nil {
+			return err
+		}
+		c.Users[email] = true
+	}
+	return save()
+}
+
+func RemoveService(i int) error {
+	c, save, err := ReadWriteConfig()
+	if err != nil {
+		return err
+	}
+
+	c.Services = append(c.Services[:i], c.Services[i+1:]...)
+	return save()
+}
+
+func RemoveAccount(i int, s int) error {
+	c, save, err := ReadWriteConfig()
+	if err != nil {
+		return err
+	}
+
+	c.Services[s].Accounts = append(c.Services[s].Accounts[:i], c.Services[s].Accounts[i+1:]...)
+	return save()
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+func NewServiceUpdater(previousName string) (*ServiceUpdater, error) {
+	c, save, err := ReadWriteConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	return &ServiceUpdater{
+		c:            c,
+		previousName: previousName,
+		save:         save,
+	}, nil
+}
+
+type ServiceUpdater struct {
+	previousName                                    string
+	name, clientID, clientSecret, authURL, tokenURL *string
+	scopes                                          *[]string
+	domains                                         *[]*Domain
+	save                                            func() error
+	c                                               *Config
+}
+
+func (u *ServiceUpdater) Commit() error {
+	var s *Service
+
+	if u.previousName == "" {
+		s = &Service{
+			OauthServiceCreds: &OauthServiceCreds{},
+		}
+		u.c.Services = append(u.c.Services, s)
+	} else {
+		for _, other := range u.c.Services {
+			if other.ServiceName == u.previousName {
+				s = other
+			}
+		}
+	}
+	if s == nil {
+		return errors.New("Unable to find service, was its name changed while editing?")
+	}
+
+	if u.name != nil {
+		s.ServiceName = *u.name
+	}
+	if u.clientID != nil {
+		s.OauthServiceCreds.ClientID = *u.clientID
+	}
+	if u.clientSecret != nil {
+		s.OauthServiceCreds.ClientSecret = *u.clientSecret
+	}
+	if u.authURL != nil {
+		s.OauthServiceCreds.AuthURL = *u.authURL
+	}
+	if u.tokenURL != nil {
+		s.OauthServiceCreds.TokenURL = *u.tokenURL
+	}
+	if u.scopes != nil {
+		s.OauthServiceCreds.Scopes = *u.scopes
+	}
+	if u.domains != nil {
+		s.Domains = u.domains
+	}
+
+	return u.save()
+}
+
+func (u *ServiceUpdater) Name(name string) error {
+	err := verifyName(name)
+	if err != nil {
+		return err
+	}
+
+	if name != u.previousName {
+		for _, other := range u.c.Services {
+			if name == other.ServiceName {
+				return errors.New("That service name is already in use.  Choose another.")
```

Comment:
general idea about all error strings in this file: I don't know how and where are these errors are going to be surfaced for the users. If it is as part of the command line interface, then they look fine. But you might want to consider adding more context to these by including the actual service name in this case or the other variables inside the error message.

Obviously you are more familiar than me in this code base and I will leave this up to you to decide. :)

Reply:
this file is used in both terminal file and ui file, well... logically UI should never show a page with these error messages displayed (mostly for debug use); for terminal, yes it would show the users. I will add some information. Thanks!